src/nodes.js

```diff
@@ -143,10 +143,55 @@ var registerChild = function(parent, key, child) {
 };
 
 
+/**
+ * Cleans a key map, nulling any nodes that are no longer in the DOM.
+ * @param {!Node} node
+ */
+var cleanKeyMap = function(node) {
+  var data = getData(node);
+  var keyMap = data.keyMap;
+  
+  if (!keyMap) {
+    return;
+  }
+
+  for (var key in keyMap) {
+    if (!keyMap[key].parentNode) {
+      delete keyMap[key];
+    }
+  }
+
+  data.keyMapValid = true;
+};
+
+
+/**
+ * Marks the key map for a given Element as invalid. It will be cleaned the
+ * next time it is needed.
+ * @param {!Node} node
+ */
+var invalidateKeyMap = function(node) {
+  getData(node).keyMapValid = false;
+};
+
+
+/**
+ * @param {!Node} node
+ * @return {boolean} True if the keyMap for the given Node is valid, false
+ *    otherwise.
+ */
+var isKeyMapValid = function(node) {
```

Comment:
Didn't we get [rid](https://github.com/google/incremental-dom/commit/f419aa0717c88c6960f591576ba71eab88f4f540) of `NodeData` getters?